### PR TITLE
Make the source directory useable as a CMake library, plus ESP32 related cleanup

### PR DESCRIPTION
**Pull request by igrr**
* `source` directory can now be used as a CMake library
  (Closes https://github.com/wasm3/wasm3/issues/29)
* esp32-idf example is refactored to include the CMake library
* esp32-pio example is refactored to use a "PIO library". Some time later people more familiar with PIO may need to tidy it up, specifically find a way to remove the symlink. Also this library can be used by other platforms which are based on PIO, so probably needs to be moved into some common directory?
* Xtensa architecture is now recognized (previously the warning was silenced with the `-w` flag).
* esp32-idf example now places all the interpreter into the IRAM, using a linker fragment file.

### CODE REVIEW

File: source/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    "m3_emit.c"
+    "m3_env.c"
+    "m3_exec.c"
+    "m3_info.c"
+    "m3_module.c"
+    "m3_optimize.c"
+    "m3_parse.c"
+)
+
+add_library(libwasm3 STATIC ${srcs})
```
Comment (igrr) [inline source/CMakeLists.txt]: @vshymanskyy  I'm not sure if this should be `wasm3` or `m3`, please advise.

Note that at the moment the static library is called `liblibwasm3.a`... The reason I didn't use `wasm3` as the target name, is that there is already a `wasm3` target defined in the main cmakelists file.

**Comment by igrr** [comment]
Looks like this has broken the wasi tests, will take a look...

File: source/CMakeLists.txt (hunk 1 above)
Comment (vshymanskyy) [reply source/CMakeLists.txt]: yes m3 is used as file prefix ;) we can include all of the `m3_*.c` files safely

Comment (igrr) [reply source/CMakeLists.txt]: so `libm3.a` for the library name and `m3` for the CMake target name is okay?

**Comment by vshymanskyy** [comment]
Should be ok!

**Comment by vshymanskyy** [comment]
@igrr Did you experience performance improvement  after putting m3 to IRAM?
Also if wasm3 is take too much IRAM, we can move just the opcodes (functions, that start with `op_`).

File: source/CMakeLists.txt (hunk 1 above)
Comment (vshymanskyy) [reply source/CMakeLists.txt]: Should be ok!

**Comment by vshymanskyy** [comment]
@igrr FYI, we no longer need the `d_m3SkipStackCheck` flag.

**Comment by vshymanskyy** [comment]
@igrr i think that failing WASI tests may be due to `d_m3HasWASI` flag, which controls WASI compilation on the source level.

**Comment by igrr** [comment]
Thanks, I'll fix this up soon. Sorry for the pause.

**Comment by vshymanskyy** [comment]
@igrr your PR is getting more importance here.
Would love if we could merge it in sooner than later! ;)

**Comment by vshymanskyy** [comment]
@igrr as for `build_wasi` failure, I can investigate it after the merge.
Even if it doesn't work at all, we can create a separate project for this.

**Comment by vshymanskyy** [comment]
BTW, PIO library is actually available now: https://github.com/wasm3/wasm3-arduino

File: CMakeLists.txt [hunk 2]
```diff
@@ -36,11 +36,6 @@ if(EMSCRIPTEN)
   set(APP_DIR            "platforms/emscripten")
 endif()
 
-if(WASIENV)
-  set(CMAKE_C_COMPILER   "wasicc")
-  set(CMAKE_CXX_COMPILER "wasic++")
-endif()
-
```
Comment (igrr) [inline CMakeLists.txt]: Now relying on `wasimake` to set the variables

File: platforms/esp32-idf/main/CMakeLists.txt [hunk 3]
```diff
@@ -1,11 +1,26 @@
...
+if (idf_ver STREQUAL "4.0")
+    # IDF v4.0 links apps with -nostdlib, so need to explicitly list the dependencies.
+
+    add_library(m3_deps INTERFACE)
+    target_link_libraries(m3_deps INTERFACE c m gcc)
+    target_link_libraries(${COMPONENT_TARGET} PRIVATE m3 m3_deps)
+else()
+    # For IDF v4.1 and later, no tricks required.
 
-target_compile_options(${COMPONENT_LIB} PRIVATE -Wno-error -O3 -DESP32 -Dd_m3LogOutput=false)
```
Comment (igrr) [inline platforms/esp32-idf/main/CMakeLists.txt]: have lost `-DESP32 -Dd_m3LogOutput=false`, let me bring this back...

File: platforms/esp32-idf/main/linker.lf [hunk 4]
```diff
@@ -0,0 +1,5 @@
+[mapping:wasm3]
+archive: liblibwasm3.a
+entries:
+    * (noflash_text)
```
Comment (igrr) [inline platforms/esp32-idf/main/linker.lf]: Which source files would you recommend to put into IRAM for performance?

File: platforms/esp32-pio/lib/m3/library.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
+    "build" : {
+        "flags": "-Dd_m3LogOutput=false -Dd_m3SkipStackCheck -O3 -w -Wfatal-errors"
```
Comment (igrr) [inline platforms/esp32-pio/lib/m3/library.json]: accidentally added `-Dd_m3SkipStackCheck` here, and lost some `Wno` flags...

**Comment by igrr** [comment]
> BTW, PIO library is actually available now: https://github.com/wasm3/wasm3-arduino

So should we remove the `esp32-pio` build here and assume that PIO users will rely on PIO+Arduino, not PIO+IDF? (I'm okay with that, less options to maintain...)

**Comment by vshymanskyy** [comment]
That library should be usable for PIO+IDF as well (but need to check).
Let's leave it as is, we'll delete it as part of some future cleanup/restructuring.

**Comment by igrr** [comment]
@vshymanskyy Okay, looks like this is ready then.

**Comment by vshymanskyy** [comment]
@igrr this is Perfect. apart from a few flags, that I'll fix later on :)
Thanks much!

**Pull request merged**